### ISSUE TITLE
Server crash when executing very long select with lots of where conditions (possible plan buffer overflow?)

### ISSUE THREAD

**Issue by fperana** [FirebirdSQL/firebird#7585]
Today I've had two server crashes after submitting a very long select statement with lots of conditions in the where clause.
This query was the result of a programmatically-built set of "or" conditions, which were mistakenly missing constraints, hence the large query.
When building the plan for this query, the server crashed.

Trying to track down the issue, I've seen that a much reduced version of the original query is working fine, then adding more and more conditions I've reached a point in which firebird returned "Error: bad allocation". Just before this point, I've realized the plan string size was 32760 bytes (approaching 32KB - maybe this is the buffer size limit for plans?).

What I've also observed is that building the execution plan takes about 0.7s for the borderline-working query. By doubling the where conditions, it then takes three times (about 2.1s) to build the plan, then it fails with "Error: bad allocation". Doubling again (so 4x the working size) it takes 7.4s before giving the bad allocation error. And finally, doubling again (8x original size), it takes about 55s and this time it crashes.

It seems that firebird tries to build an execution plan even for queries far beyond its limits, failing the bad way.

FB version is 4.0.2 on Debian 11 x64.

**Comment by fperana** [comment]
Core dump of the second event, if it can be useful.

[core.001.zip](https://github.com/FirebirdSQL/firebird/files/11478498/core.001.zip)
[core.002.zip](https://github.com/FirebirdSQL/firebird/files/11478500/core.002.zip)

(this is a single zip file split into two files, you have to merge 001 and 002 into a single file and unzip this one)

**Comment by dyemanov** [comment]
The engine cannot handle unlimited number of ANDed or ORed conditions because they're processed (both compiled and executed) recursively and the crash is unavoidable at some level of recursion (due to limited stack size). Perhaps we could add some sanity check for such kind of implicit recursion (like we have 1500 items for IN -- for exactly the same reason)

**Comment by fperana** [comment]
Yes I understand this as a developer. I strongly suggest to add some sanity check before or while compiling a submitted statement, I seriously think it's not a good thing someone can crash the server only by malformed / excessive complex sql.

**Comment by asfernandes** [comment]
Very easy to say, but don't forget that we can have the combinations with different operators (not only logical).
